File: macos/SwiftShareDomain/TransferRoles.swift
import Foundation

public enum TransferActivityPhase: Equatable, Sendable {
    case preparing
    case connecting
    case awaitingApproval
    case transferring
    case verifying
    case committing
    case completed
    case rejected
    case cancelled
    case failed
}

public struct TransferSessionEvent: Equatable, Sendable {
    public let phase: TransferActivityPhase
    public let transferredBytes: UInt64
    public let verifiedBytes: UInt64
    public let totalBytes: UInt64
    public let completedItems: Int
    public let totalItems: Int

    public init(
        phase: TransferActivityPhase,
        transferredBytes: UInt64 = 0,
        verifiedBytes: UInt64 = 0,
        totalBytes: UInt64 = 0,
        completedItems: Int = 0,
        totalItems: Int = 1
    ) {
        self.phase = phase
        self.transferredBytes = transferredBytes
        self.verifiedBytes = verifiedBytes
        self.totalBytes = totalBytes
        self.completedItems = completedItems
        self.totalItems = totalItems
    }
}

public struct TransferFailure: Equatable, Sendable {
    public let category: TransferErrorCategory
    public let code: String

    public init(category: TransferErrorCategory, code: String) {
        self.category = category
        self.code = code
    }
}

public struct RoleTransferOutcome: Equatable, Sendable {
    public let result: TransferTerminalResult
    public let committedArtifacts: [CommittedArtifact]
    public let failure: TransferFailure?

    public init(
        result: TransferTerminalResult,
        committedArtifacts: [CommittedArtifact] = [],
        failure: TransferFailure? = nil
    ) {
        self.result = result
        self.committedArtifacts = committedArtifacts
        self.failure = failure
    }
}

public struct TransferTimeoutPolicy: Equatable, Sendable {
    public let connect: Duration
    public let approval: Duration
    public let idle: Duration

    public init(
        connect: Duration = .seconds(15),
        approval: Duration = .seconds(120),
        idle: Duration = .seconds(60)
    ) {
        self.connect = connect
        self.approval = approval
        self.idle = idle
    }
}

public protocol TransferEventSink: Sendable {
    func emit(_ event: TransferSessionEvent) async
}

public protocol TransferCancellationChecking: Sendable {
    func isCancelled() async -> Bool
}

public protocol TransferDeadlineScheduling: Sendable {
    func run<T: Sendable>(
        timeout: Duration,
        operation: @escaping @Sendable () async throws -> T
    ) async throws -> T
}

public struct InboundTransferContext: Sendable {
    public let request: TransferSessionRequest
    public let authenticatedPeerID: String

    public init(request: TransferSessionRequest, authenticatedPeerID: String) {
        self.request = request
        self.authenticatedPeerID = authenticatedPeerID
    }
}

public protocol IncomingApprovalGateway: Sendable {
    func requestApproval(peer: PairedPeer, request: TransferSessionRequest) async throws -> Bool
}

public protocol PeerApprovalPolicyResolving: Sendable {
    func policy(for peer: PairedPeer) async throws -> PeerApprovalPolicy
}

public struct RequireApprovalPolicyResolver: PeerApprovalPolicyResolving {
    public init() {}
    public func policy(for peer: PairedPeer) async throws -> PeerApprovalPolicy { .requireApproval }
}

public protocol InboundTransferReceiving: Sendable {
    func receive(
        peer: PairedPeer,
        request: TransferSessionRequest,
        events: any TransferEventSink,
        cancellation: any TransferCancellationChecking,
        deadlineScheduler: any TransferDeadlineScheduling,
        idleTimeout: Duration
    ) async throws -> CommittedArtifact
}

public struct InboundTransferSessionCoordinator: Sendable {
    private let peerDirectory: any PeerDirectory
    private let approvalPolicyResolver: any PeerApprovalPolicyResolving
    private let approvalGateway: any IncomingApprovalGateway
    private let receiver: any InboundTransferReceiving
    private let deadlineScheduler: any TransferDeadlineScheduling
    private let cancellation: any TransferCancellationChecking
    private let events: any TransferEventSink
    private let timeouts: TransferTimeoutPolicy

    public init(
        peerDirectory: any PeerDirectory,
        approvalPolicyResolver: any PeerApprovalPolicyResolving = RequireApprovalPolicyResolver(),
        approvalGateway: any IncomingApprovalGateway,
        receiver: any InboundTransferReceiving,
        deadlineScheduler: any TransferDeadlineScheduling,
        cancellation: any TransferCancellationChecking,
        events: any TransferEventSink,
        timeouts: TransferTimeoutPolicy = TransferTimeoutPolicy()
    ) {
        self.peerDirectory = peerDirectory
        self.approvalPolicyResolver = approvalPolicyResolver
        self.approvalGateway = approvalGateway
        self.receiver = receiver
        self.deadlineScheduler = deadlineScheduler
        self.cancellation = cancellation
        self.events = events
        self.timeouts = timeouts
    }

    public func execute(_ context: InboundTransferContext) async -> RoleTransferOutcome {
        let request = context.request
        do {
            guard let peer = try await peerDirectory.pairedPeer(id: context.authenticatedPeerID) else {
                return await terminal(.rejected, phase: .rejected, request: request)
            }
            await events.emit(event(.awaitingApproval, request: request))
            let approved = switch try await approvalPolicyResolver.policy(for: peer) {
            case .autoAccept: true
            case .requireApproval:
                try await deadlineScheduler.run(timeout: timeouts.approval) {
                    try await approvalGateway.requestApproval(peer: peer, request: request)
                }
            }
            guard approved else { return await terminal(.rejected, phase: .rejected, request: request) }
            guard !(await cancellation.isCancelled()) else {
                return await terminal(.cancelled, phase: .cancelled, request: request)
            }
            await events.emit(event(.transferring, request: request))
            let artifact = try await receiver.receive(
                peer: peer,
                request: request,
                events: events,
                cancellation: cancellation,
                deadlineScheduler: deadlineScheduler,
                idleTimeout: timeouts.idle
            )
            await events.emit(event(.completed, request: request, verifiedBytes: request.totalBytes))
            return RoleTransferOutcome(result: .completed, committedArtifacts: [artifact])
        } catch let error as TransferExecutionError {
            await events.emit(event(.failed, request: request))
            return RoleTransferOutcome(
                result: .failed,
                failure: TransferFailure(category: error.category, code: error.code)
            )
        } catch {
            await events.emit(event(.failed, request: request))
            return RoleTransferOutcome(
                result: .failed,
                failure: TransferFailure(category: .protocol, code: "unexpected_failure")
            )
        }
    }

    private func terminal(
        _ result: TransferTerminalResult,
        phase: TransferActivityPhase,
        request: TransferSessionRequest
    ) async -> RoleTransferOutcome {
        await events.emit(event(phase, request: request))
        return RoleTransferOutcome(result: result)
    }
}

public struct TransferExecutionError: Error, Equatable, Sendable {
    public let category: TransferErrorCategory
    public let code: String

    public init(category: TransferErrorCategory, code: String) {
        self.category = category
        self.code = code
    }
}

private func event(
    _ phase: TransferActivityPhase,
    request: TransferSessionRequest,
    verifiedBytes: UInt64 = 0
) -> TransferSessionEvent {
    TransferSessionEvent(
        phase: phase,
        verifiedBytes: verifiedBytes,
        totalBytes: request.totalBytes,
        completedItems: phase == .completed ? request.payloads.count : 0,
        totalItems: request.payloads.count
    )
}

private extension TransferSessionRequest {
    var totalBytes: UInt64 { payloads.reduce(0) { $0 + $1.byteCount } }
}

File: android/domain/src/main/kotlin/dev/swiftshare/domain/TransferRoles.kt
package dev.swiftshare.domain

enum class TransferActivityPhase {
    PREPARING,
    CONNECTING,
    AWAITING_APPROVAL,
    TRANSFERRING,
    VERIFYING,
    COMMITTING,
    COMPLETED,
    REJECTED,
    CANCELLED,
    FAILED,
}

data class TransferSessionEvent(
    val phase: TransferActivityPhase,
    val transferredBytes: Long = 0,
    val verifiedBytes: Long = 0,
    val totalBytes: Long = 0,
    val completedItems: Int = 0,
    val totalItems: Int = 1,
    val currentItemName: String = "",
)

data class TransferFailure(
    val category: TransferErrorCategory,
    val code: String,
)

data class RoleTransferOutcome(
    val result: TransferTerminalResult,
    val committedArtifacts: List<CommittedArtifact> = emptyList(),
    val failure: TransferFailure? = null,
)

data class TransferTimeoutPolicy(
    val connectMillis: Long = 15_000,
    val approvalMillis: Long = 120_000,
    val idleMillis: Long = 60_000,
)

fun interface TransferEventSink {
    suspend fun emit(event: TransferSessionEvent)
}

fun interface TransferCancellationChecking {
    suspend fun isCancelled(): Boolean
}

interface TransferDeadlineScheduler {
    suspend fun <T> run(timeoutMillis: Long, operation: suspend () -> T): T
}

data class OutboundConnectionContext(
    val request: TransferSessionRequest,
    val localDevice: DeviceIdentity,
    val peer: PairedPeer,
    val endpoint: PeerEndpoint,
)

data class RemoteTransferResult(
    val status: WireTerminalStatus,
    val committedArtifactId: String = "",
    val failure: TransferFailure? = null,
)

interface OutboundTransferConnection {
    suspend fun awaitApproval(): TransferApproval

    suspend fun stream(
        events: TransferEventSink,
        cancellation: TransferCancellationChecking,
        deadlineScheduler: TransferDeadlineScheduler,
        idleTimeoutMillis: Long,
    ): RemoteTransferResult

    suspend fun cancel(reasonCode: String)
}

fun interface OutboundTransferConnector {
    suspend fun connect(context: OutboundConnectionContext): OutboundTransferConnection
}

class OutboundTransferSessionCoordinator(
    private val identityStore: IdentityStore,
    private val peerDirectory: PeerDirectory,
    private val endpointResolver: PeerEndpointResolver,
    private val connector: OutboundTransferConnector,
    private val deadlineScheduler: TransferDeadlineScheduler,
    private val cancellation: TransferCancellationChecking,
    private val events: TransferEventSink,
    private val timeouts: TransferTimeoutPolicy = TransferTimeoutPolicy(),
) {
    suspend fun execute(request: TransferSessionRequest): RoleTransferOutcome {
        return try {
            events.emit(event(TransferActivityPhase.PREPARING, request))
            val peer = peerDirectory.pairedPeer(request.peerId)
                ?: return terminal(TransferTerminalResult.REJECTED, TransferActivityPhase.REJECTED, request)
            val localDevice = identityStore.localDeviceIdentity()
            val endpoint = endpointResolver.endpoint(peer)
            events.emit(event(TransferActivityPhase.CONNECTING, request))
            val connection = deadlineScheduler.run(timeouts.connectMillis) {
                connector.connect(OutboundConnectionContext(request, localDevice, peer, endpoint))
            }
            if (cancellation.isCancelled()) {
                connection.cancel("user_cancelled")
                return terminal(TransferTerminalResult.CANCELLED, TransferActivityPhase.CANCELLED, request)
            }
            events.emit(event(TransferActivityPhase.AWAITING_APPROVAL, request))
            val approval = deadlineScheduler.run(timeouts.approvalMillis) { connection.awaitApproval() }
            if (!approval.accepted) {
                return terminal(TransferTerminalResult.REJECTED, TransferActivityPhase.REJECTED, request)
            }
            events.emit(event(TransferActivityPhase.TRANSFERRING, request))
            val remote = connection.stream(events, cancellation, deadlineScheduler, timeouts.idleMillis)
            val status = remote.status.toDomainResult()
            val phase = status.toActivityPhase()
            events.emit(event(phase, request, verified = request.totalBytes()))
            RoleTransferOutcome(
                result = status,
                committedArtifacts = if (status == TransferTerminalResult.COMPLETED && remote.committedArtifactId.isNotEmpty()) {
                    listOf(CommittedArtifact(remote.committedArtifactId))
                } else emptyList(),
                failure = remote.failure,
            )
        } catch (error: TransferExecutionException) {
            events.emit(event(TransferActivityPhase.FAILED, request))
            RoleTransferOutcome(
                result = TransferTerminalResult.FAILED,
                failure = TransferFailure(error.category, error.errorCode),
            )
        } catch (_: Exception) {
            events.emit(event(TransferActivityPhase.FAILED, request))
            RoleTransferOutcome(
                result = TransferTerminalResult.FAILED,
                failure = TransferFailure(TransferErrorCategory.PROTOCOL, "unexpected_failure"),
            )
        }
    }

    private suspend fun terminal(
        result: TransferTerminalResult,
        phase: TransferActivityPhase,
        request: TransferSessionRequest,
    ): RoleTransferOutcome {
        events.emit(event(phase, request))
        return RoleTransferOutcome(result)
    }
}

class TransferExecutionException(
    val category: TransferErrorCategory,
    val errorCode: String,
) : Exception(errorCode)

private fun TransferSessionRequest.totalBytes(): Long = payloads.sumOf { it.byteCount }

private fun event(
    phase: TransferActivityPhase,
    request: TransferSessionRequest,
    verified: Long = 0,
): TransferSessionEvent = TransferSessionEvent(
    phase = phase,
    verifiedBytes = verified,
    totalBytes = request.totalBytes(),
    completedItems = if (phase == TransferActivityPhase.COMPLETED) request.payloads.size else 0,
    totalItems = request.payloads.size,
)

private fun WireTerminalStatus.toDomainResult(): TransferTerminalResult = when (this) {
    WireTerminalStatus.COMPLETED -> TransferTerminalResult.COMPLETED
    WireTerminalStatus.REJECTED -> TransferTerminalResult.REJECTED
    WireTerminalStatus.CANCELLED -> TransferTerminalResult.CANCELLED
    WireTerminalStatus.FAILED -> TransferTerminalResult.FAILED
}

private fun TransferTerminalResult.toActivityPhase(): TransferActivityPhase = when (this) {
    TransferTerminalResult.COMPLETED -> TransferActivityPhase.COMPLETED
    TransferTerminalResult.REJECTED -> TransferActivityPhase.REJECTED
    TransferTerminalResult.CANCELLED -> TransferActivityPhase.CANCELLED
    TransferTerminalResult.FAILED -> TransferActivityPhase.FAILED
}
